src/bus/driver.c

```diff
@@ -1341,28 +1357,11 @@ static int driver_method_get_connection_unix_process_id(Peer *peer, const char *
         return 0;
 }
 
-static void driver_append_connection_credentials(CDVar *v, Bus *bus, Peer *peer) {
-        const char *seclabel;
-        size_t n_seclabel, n_gids;
-        gid_t *gids;
-        uid_t uid;
-        pid_t pid;
+static void driver_write_credentials_fields(CDVar *v, const char *seclabel, size_t n_seclabel, gid_t *gids, size_t n_gids, uid_t uid, pid_t pid, int pid_fd, size_t *n_fds) {
 
-        if (peer) {
-                uid = peer->user->uid;
-                pid = peer->pid;
-                gids = peer->gids;
-                n_gids = peer->n_gids;
-                seclabel = peer->seclabel;
-                n_seclabel = peer->n_seclabel;
-        } else {
-                uid = bus->user->uid;
-                pid = bus->pid;
-                gids = bus->gids;
-                n_gids = bus->n_gids;
-                seclabel = bus->seclabel;
-                n_seclabel = bus->n_seclabel;
-        }
+        assert(v);
+        assert(gids || n_gids == 0);
+        assert(n_fds || pid_fd < 0);
 
         c_dvar_write(v, "[{s<u>}{s<u>}",
                      "UnixUserID", c_dvar_type_u, uid,
```

Comment:
we never get a negative pid here - I'll add an assertion to make it clear. Both callers of this helper bail out earlier if the pid cannot be resolved, because it means the service is gone, so we don't have any data to return

Reply:
Yeah, i missed the `... || pid < 0` part of the conditional in the caller. But I think that is wrong. You suddenly stop reporting information for a peer just because its _original_ process is gone. I don't think that is ok.

I think it is fine to drop the PID, because it can be recycled and thus actually poses a security risk. But data that is just possibly outdated, but still represents what the peer was at the time of connect() should still be included. And even if you happen to drop all the information, then the function should still return success, but with an empty array. Otherwise, you will make it unnecessarily hard for future extensions, because clients suddenly need to deal with error-returns *AND* missing fields, rather than just missing fields.

Or what exactly is the point of returning an error?